### ISSUE TITLE
Tailcall optimization issue 

### ISSUE THREAD

**Issue by ncave** [fable-compiler/Fable#681]
This works on dotnet and Fable 0.7.35, but not on Fable 0.7.40:
```
let rec test x f = 
    match x with
    | [] -> f x
    | h::t -> test t (f << id)

[<EntryPoint>]
let main argv =
    let a = test [1;2;3] id
    printfn "%A" a
    0
```
Here is how it compiles on 0.7.35 (works correctly):
```
function test(x, f) {
  if (x.tail != null) {
    return test(x.tail, $var1 => f((x_1 => x_1)($var1)));
  } else {
    return f(x);
  }
}
```
Here is how it compiles on 0.7.40 (runtime error):
```
function test(x, f) {
  test: while (true) {
    if (x.tail != null) {
      x = x.tail;
      f = $var1 => f((x_1 => x_1)($var1));
      continue test;
    } else {
      return f(x);
    }
  }
}
```

**Comment by alfonsogarciacaro** [comment]
So generalizing this I guess we should avoid tailcall optimizations when there're function arguments as they may include delayed references to themselves or other arguments that would break when changing the recursion into a loop.

It's a bit of a pity because that means the continuation passing style cannot be optimized, but I cannot think of another solution at the moment. Please let me know if you've any suggestion.

**Comment by Pauan** [comment]
@alfonsogarciacaro Can't it store the previous value of `f` in a `const`, like this?

```
function test(x, f) {
  test: while (true) {
    if (x.tail != null) {
      x = x.tail;
      const f1 = f;
      f = $var1 => f1((x_1 => x_1)($var1));
      continue test;
    } else {
      return f(x);
    }
  }
}
```

**Comment by alfonsogarciacaro** [comment]
Good point, @Pauan! I didn't know `const` could hold different values when defined multiple times in the same loop. And it seems [Babel can match this with `var` too](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=env&targets=&browsers=&builtIns=false&code=function%20test(x%2C%20f)%20%7B%0D%0A%20%20test%3A%20while%20(true)%20%7B%0D%0A%20%20%20%20if%20(x.tail%20!%3D%20null)%20%7B%0D%0A%20%20%20%20%20%20x%20%3D%20x.tail%3B%0D%0A%20%20%20%20%20%20const%20f1%20%3D%20f%3B%0D%0A%20%20%20%20%20%20f%20%3D%20%24var1%20%3D%3E%20f1((x_1%20%3D%3E%20x_1)(%24var1))%3B%0D%0A%20%20%20%20%20%20continue%20test%3B%0D%0A%20%20%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%20%20return%20f(x)%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%0D%0A%7D) 👍 

I will fix it, thanks again for the advice!

**Comment by Pauan** [comment]
@alfonsogarciacaro Yeah, it's because `var` has function scoping, but `let` and `const` both have block scoping. So every time it evaluates the block, it will create a fresh binding for the `const`, and that fresh binding can be captured by the closure.

**Comment by alfonsogarciacaro** [comment]
This should have been fixed in fable-compiler 0.7.46. Please give it a try :)
